core-bundle/src/Routing/RouteProvider.php

```diff
@@ -491,11 +490,12 @@ private function findPages(array $candidates): array
             return [];
         }
 
+        /** @var array<PageModel> */
```

Comment:
Shouldn't we add `@mixin<T>` and `@return array<T>` (or some other generics magic) to the `Model` class so PhpStan knows what we are returning here?

Reply:
That's a tough one. We would imo need to add generics to the `Collection`, the `Adapter` and the `Model` for this to work. And then we will end with a lot of `@phpstan-param`, `@phpstan-return`, `@phpstan-var`... declarations so that our IDEs are still able to guess the type. 

Everything would look a bit like this I guess: https://phpstan.org/r/15377a79-1be1-4f2d-999f-f1ae9dbe37f0

Reply:
Do you want to try?

Reply:
Challenge accepted. Are you ok with all those tags?

Reply:
I had a deep dive into the topic. 

Most of the things could *theoretically* be solved by having a `Collection<T of Model>` + `static`/`Collection<static>` return type annotations in the model base class and forced return overwrites for all of the `findBy*` methods because of the dynamic `Model` vs `Collection` vs `Array` return type nature of `find()`. 

Unfortunately, in reality this doesn't work as expected because we're unfortunately using `__callStatic` and PHPStan then is not able to resolve the types.

It would possibly work to create our own extension, but well...

Reply:
Feel free to modify https://github.com/contao/phpstan 😎

Reply:
> Feel free to modify https://github.com/contao/phpstan sunglasses 

If I only knew how. :smile:

Reply:
I have created a follow-up issue here: #1940